File: src/main/scala/com/nekokittygames/Thaumic/Tinkerer/common/blocks/BlockFunnel.scala
package com.nekokittygames.Thaumic.Tinkerer.common.blocks

import com.nekokittygames.Thaumic.Tinkerer.common.libs.LibNames
import com.nekokittygames.Thaumic.Tinkerer.common.tiles.TileFunnel
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.block.properties.PropertyBool
import net.minecraft.block.state.{BlockState, IBlockState}
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Blocks
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.{BlockPos, EnumFacing, EnumWorldBlockLayer}
import net.minecraft.world.{IBlockAccess, World}
import net.minecraftforge.fml.relauncher.SideOnly

import scala.util.Random

/**
 * Created by Katrina on 18/05/2015.
 */
object BlockFunnel extends {
  val JAR:PropertyBool=PropertyBool.create("jar")
} with ModBlockContainer(Material.iron)
{
  setHardness(3.0F)
  setResistance(8.0f)
  setStepSound(Block.soundTypeStone)
  setBlockBounds(0f, 0f, 0f, 1f, 1f / 8f, 1f)
  setDefaultState(getDefaultState.withProperty(JAR, false))
  setUnlocalizedName(LibNames.FUNNEL)
  val random: Random = new Random()

  override def createBlockState(): BlockState = new BlockState(this,JAR)


  override def getActualState(state: IBlockState, worldIn: IBlockAccess, pos: BlockPos): IBlockState =
    {
      val te:TileEntity=worldIn.getTileEntity(pos)
      if(te.isInstanceOf[TileFunnel])
        {
          if(te.asInstanceOf[TileFunnel].inventory==null)
            state.withProperty(JAR,false)
          else
            state.withProperty(JAR,true)
        }
      else
        state.withProperty(JAR,false)
    }

  override def getMetaFromState(state: IBlockState): Int = {0}


  override def getStateFromMeta(meta: Int): IBlockState = {getDefaultState}

  @SideOnly(net.minecraftforge.fml.relauncher.Side.CLIENT)
  override def getBlockLayer: EnumWorldBlockLayer = EnumWorldBlockLayer.TRANSLUCENT

  override def isOpaqueCube(): Boolean = false

  override def isFullCube(): Boolean = false


  override def getTileClass: Class[_ <: TileEntity] = classOf[TileFunnel]

  override def canPlaceBlockAt(worldIn: World, pos: BlockPos): Boolean = {
    return worldIn.getBlockState(pos.down(1)).getBlock == Blocks.hopper
  }

  override def onNeighborChange(world: IBlockAccess, pos: BlockPos, neighbor: BlockPos): Unit = {
    if (world.getBlockState(pos.down(1)).getBlock != Blocks.hopper) {
      dropBlockAsItem(world.asInstanceOf[World], pos, world.getBlockState(pos), 0)
      world.asInstanceOf[World].setBlockToAir(pos)
    }
  }

  override def onBlockActivated(worldIn: World, pos: BlockPos, state: IBlockState, playerIn: EntityPlayer, side: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): Boolean = {
    val funnel = worldIn.getTileEntity(pos).asInstanceOf[TileFunnel]
    val itemStack = funnel.inventory

    if (itemStack == null) {
      val playrStack = playerIn.getCurrentEquippedItem
      if (playrStack != null && funnel.canInsertItem(0, playrStack, EnumFacing.DOWN)) {
        funnel.setInventorySlotContents(0, playrStack.splitStack(1))
        if (playrStack.stackSize <= 0) {
          playerIn.inventory.setInventorySlotContents(playerIn.inventory.currentItem, null)
        }
        funnel.markDirty()
        return true
      }
    }
    else {
      if (!playerIn.inventory.addItemStackToInventory(itemStack))
        playerIn.dropPlayerItemWithRandomChoice(itemStack, false)
      funnel.setInventorySlotContents(0, null)
      funnel.markDirty()
      return true
    }

    return false
  }
}

File: src/main/java/com/nekokittygames/thaumictinkerer/common/blocks/BlockFunnel.java
package com.nekokittygames.thaumictinkerer.common.blocks;

import com.nekokittygames.thaumictinkerer.common.libs.LibBlockNames;
import com.nekokittygames.thaumictinkerer.common.tileentity.TileEntityFunnel;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.block.properties.PropertyBool;
import net.minecraft.block.state.BlockFaceShape;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;

public class BlockFunnel extends TTTileEntity<TileEntityFunnel> {

    public static final PropertyBool JAR= PropertyBool.create("jar");

    public BlockFunnel() {
        super(LibBlockNames.FUNNEL, Material.ROCK,true);
        setHardness(3.0F);
        setResistance(8.0f);
        setDefaultState(this.getBlockState().getBaseState().withProperty(JAR,false));

    }

    @Override
    public BlockRenderLayer getBlockLayer() {
        return BlockRenderLayer.TRANSLUCENT;
    }

    @Override
    public boolean isTranslucent(IBlockState state) {
        return true;
    }

    @Override
    public boolean isOpaqueCube(IBlockState state) {
        return false;
    }

    @Override
    protected BlockStateContainer createBlockState() {
        return new BlockStateContainer(this,JAR);
    }

    @Override
    public IBlockState getActualState(IBlockState state, IBlockAccess worldIn, BlockPos pos) {
        TileEntity te= (TileEntity) worldIn.getTileEntity(pos);
        if(te instanceof TileEntityFunnel)
        {
            TileEntityFunnel funnel=(TileEntityFunnel)te;
            if(funnel.getInventory().getStackInSlot(0)== ItemStack.EMPTY)
                return state.withProperty(JAR,false);
            else
                return state.withProperty(JAR,true);
        }
        return state.withProperty(JAR,false);
    }

    @Override
    public BlockFaceShape getBlockFaceShape(IBlockAccess worldIn, IBlockState state, BlockPos pos, EnumFacing face) {
        return BlockFaceShape.UNDEFINED;
    }
    @Override
    public int getMetaFromState(IBlockState state) {
        return 0;
    }

    @Override
    public IBlockState getStateFromMeta(int meta) {
        return this.blockState.getBaseState().withProperty(JAR,false);
    }

    @Override
    public boolean canPlaceBlockAt(World worldIn, BlockPos pos) {
        return super.canPlaceBlockAt(worldIn, pos) && worldIn.getBlockState(pos.down()).getBlock()== Blocks.HOPPER;
    }

    @Override
    public TileEntity createTileEntity(World world, IBlockState state) {
        return new TileEntityFunnel();
    }


    @Override
    public boolean onBlockActivated(World worldIn, BlockPos pos, IBlockState state, EntityPlayer playerIn, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        TileEntity te=worldIn.getTileEntity(pos);
        if(te instanceof TileEntityFunnel)
        {
            TileEntityFunnel funnel= (TileEntityFunnel) te;
            ItemStack stack=funnel.getInventory().getStackInSlot(0);
            if(stack==ItemStack.EMPTY)
            {
                ItemStack playerStack=playerIn.getHeldItem(hand);
                if(funnel.isItemValidForSlot(0,playerStack))
                {
                    funnel.getInventory().insertItem(0,playerStack.copy(),false);
                    playerStack.setCount(playerStack.getCount()-1);
                    if (playerStack.isEmpty()) {
                        playerIn.inventory.setInventorySlotContents(playerIn.inventory.currentItem, ItemStack.EMPTY);
                    }
                    funnel.markDirty();
                    return true;
                }
            }
            else
            {
                if(!playerIn.inventory.addItemStackToInventory(stack.copy()))
                {
                    playerIn.dropItem(stack,false);

                }
                funnel.getInventory().setStackInSlot(0,ItemStack.EMPTY);
                funnel.markDirty();
                return true;
            }
        }
        return false;
    }
}
